### DISCUSSION TITLE
Support MongoDB $jsonSchema

### DISCUSSION THREAD

**Discussion by bwgjoseph** [jy95/docusaurus-json-schema-plugin#19] [Ideas]
Is it possible to support it?

See [mongodb-jsonschema](https://www.mongodb.com/docs/manual/reference/operator/query/jsonSchema/)

A sample

```json
{ $jsonSchema: <JSON Schema object> }
```

```json
{
  $jsonSchema: {
     required: [ "name", "major", "gpa", "address" ],
     properties: {
        name: {
           bsonType: "string",
           description: "must be a string and is required"
        },
        address: {
           bsonType: "object",
           required: [ "zipcode" ],
           properties: {
               "street": { bsonType: "string" },
               "zipcode": { bsonType: "string" }
           }
        }
     }
  }
}
```

If it does, I imagine I would use the same way I did before to display it

```
import JSONSchemaViewer from "@theme/JSONSchemaViewer"
import schema from "@site/static/schema/mongo.json"

<JSONSchemaViewer schema={schema} />
```

Thanks

**Comment by jy95** [comment]
Hello @bwgjoseph ,

### Short answer :  

Low priority for me atm.
A proper way to extend the best effort viewer with extensions of JSON Schema needs some reflexion first.
I would like to avoid rushing things that would leak to issues , breaking changes, ...

Why the low priority ?
* According to [ajv](https://ajv.js.org/guide/schema-language.html), Draft-07 is the most widely used version and by consequence, draft-04 is deprecated & many packages don't even care about it anymore like the [library ](https://www.npmjs.com/package/json-schema-typed) I use for the Typescript types
* Viewer main targets are Draft-07 / Draft 2019-09 / Draft 2020-12, as expressed in the readme
* There are some [backwards-incompatible changes between draft-04 & other versions](https://json-schema.org/draft-06/json-schema-release-notes.html#q-what-are-the-changes-between-draft-04-and-draft-06), such as on `exclusiveMaximum` / `exclusiveMinimum` definitions
* Discussion about that specs on MongoDB forums : https://www.mongodb.com/community/forums/t/validation-using-latest-json-schema-version/2082
* ...


### Long answer :

Open closed principle should be applied here.
MongoDB extended Draft 4 of JSON specs with 3 major differences : 
1. `$jsonSchema` is their root element
2. `exclusiveMaximum` / `exclusiveMinimum` breaking changes definitions
3. Their `bsonType` (similar to `type` but with more )

#### For point n°1

You can use the [`jsonPointer`](https://gregsdennis.github.io/Manatee.Json/usage/pointer.html) option inside resolverOptions , such as : 

```tsx
import Schema from "@site/static/schemas/examples/refs/refOnRoot.json";
import JSONSchemaViewer from "@theme/JSONSchemaViewer";

const resolverOptions = {"jsonPointer": "#/definitions/user"}

<JSONSchemaViewer schema={ Schema } resolverOptions={resolverOptions} />
```

My example [in documentation](https://jy95.github.io/docusaurus-json-schema-plugin/docs/demo-viewer/refs/ref) is a good illustration it already works.    
If you wonder why this option already exists, well, I had that issue with some of my Draft 7 files I have in one of my other projects 🤐

#### For point n°2

In my own codebase, bunch of 

```tsx
// @ts-ignore
as boolean | number | undefined
```

to just cover that version. It seems to be an overhead just to support a single version ...

#### For point n°3

The hardest as many question aroses : 
- Should I enable users to provide their own components to render their custom types ?
- Should I provide a default rendering if users don't provide a fallback one ?
- ...

So I will convert that issue into a discussion to see if other have this need that might help me reconsider my position.
Part of your request (the need for customization) is a track from which I would like other feedbacks: at least, on the internalization side of the labels I am ready ^^

PS : 1.2.0 is out. enhancements should be quite visible : feel free to share your feedback ;)

**Reply by bwgjoseph** [reply]
hey @jy95,

Appreciate the detailed response. I didn't realize that using 04 would be that unsupported (I didn't look it up previously). But I do agree with you that given this is the case, I'm not sure if supporting `mongo $jsonSchema` make any sense at all. It's really my personal use case for now.

Even if it's not supported officially, I should be able to make some small tweaks to the `mongo $jsonSchema` and "mock" it as a normal $jsonSchema and it would have shown up quite nicely (I think). I haven't tried the complicated ones.

I went ahead to test out what you said, and I did not get you wrong, is this correct? If so, I'm not sure why it is not working.

```js
import CodeBlock from '@theme/CodeBlock'
import JSONSchemaViewer from "@theme/JSONSchemaViewer"
import schema from "@site/static/schema/mongo.json"
const resolverOptions = { "jsonPointer": "#/definitions/$jsonSchema" }

## Viewer

<JSONSchemaViewer schema={schema} resolverOptions={resolverOptions} />

## Source

<details>
  <summary>Click me</summary>

<CodeBlock language="json">{JSON.stringify(schema, null, 2)}</CodeBlock>
</details>
```

```json
{
    "$ref": "#/definitions/$jsonSchema",
    "$jsonSchema": {
        "bsonType":"object",
        "required":[
           "title",
           "body"
        ],
        "properties":{
           "title":{
              "bsonType":"string",
              "description":"Title of post - Required."
           },
           "body":{
              "bsonType":"string",
              "description":"Body of post - Required."
           },
           "category":{
              "bsonType":"string",
              "description":"Category of post - Optional."
           },
           "likes":{
              "bsonType":"int",
              "description":"Post like count. Must be an integer - Optional."
           },
           "tags":{
              "bsonType":[
                 "string"
              ],
              "description":"Must be an array of strings - Optional."
           },
           "date":{
              "bsonType":"date",
              "description":"Must be a date - Optional."
           }
        }
     }
}
```

Gave me the "loading..." error again

![image](https://user-images.githubusercontent.com/37484538/219876186-dc5875e9-5ff5-4126-9c79-3db69367fb7c.png)

Thanks for looking into it

**Reply by jy95** [reply]
Obviously, your `jsonPointer` / schema is wrong - that is why you get our old pal ...
(At some point, writting a FAQ page with "Why do I get the loading screen ?" is something useful 🤣)

A workaround for the time being (not 100% bulletproof I should say) could be : 

```tsx
// Maybe in your site /src folder, define a _convertToDraft7 file to do the conversion
export default function convertToDraft7(mongoSchema): {[x: string] : any} {

    const DRAFT_07_TYPES = ["null", "integer", "number", "object", "array", "boolean", "string"]

    // TODO - based on "mongoSchema", I must do the following steps recursively :
    // 1A. If "exclusiveMinimum" === true, "minimum" becomes "exclusiveMinimum" in "mongoSchema"
    // 1B. If "exclusiveMaximum" === true, "maximum" becomes "exclusiveMaximum" in "mongoSchema"
    // 2A. If "bsonType" AND is included in DRAFT_07_TYPES, copy it in "type"
    // 2B. If "bsonType" AND is NOT included in DRAFT_07_TYPES, create {"type": "string", "format": yourBsonType}
    // 3. If "bsonType" is an array, apply for each element step 2A & 2B
    let rewritedMongoDb = {}

    return {
        "$ref": "#/definitions/mongoDB",
        "definitions": {
            "mongoDB": rewritedMongoDb
        }
    }
}

// which you can later invoke in mdx files like
import JSONSchemaViewer from "@theme/JSONSchemaViewer"
import schema from "@site/static/schema/mongo.json"
import convertToDraft7 from "@site/src/_convertToDraft7"
const resolverOptions = { "jsonPointer": "#/definitions/mongoDB" }

<JSONSchemaViewer schema={convertToDraft7(schema)} resolverOptions={resolverOptions} />
```

Of course, for the actual implementation, you can ask ChatGPT or others to do the job. 
It can do wonder when situation is cleary expressed like I tried in Typescript ;)

**Reply by jy95** [reply]
@bwgjoseph If you wonder, here is what ChatGPT generated, with my comments : 
⚠️It seems to follow what I expressed, but as usually we should be careful with what he gave us

```tsx
export default function convertToDraft7(mongoSchema): {[x: string] : any} {

    const DRAFT_07_TYPES = ["null", "integer", "number", "object", "array", "boolean", "string"]

    function convertObject(schema) {
        const convertedSchema = {}
        for (const key in schema) {
            if (key === "exclusiveMinimum" && schema[key] === true) {
                convertedSchema["exclusiveMinimum"] = schema["minimum"]
            } else if (key === "exclusiveMaximum" && schema[key] === true) {
                convertedSchema["exclusiveMaximum"] = schema["maximum"]
            } else if (key === "bsonType") {
                if (Array.isArray(schema[key])) {
                    convertedSchema["type"] = "array"
                    convertedSchema["items"] = schema[key].map(convertBsonType)
                } else {
                    const bsonType = schema[key]
                    if (DRAFT_07_TYPES.includes(bsonType)) {
                        convertedSchema["type"] = bsonType
                    } else {
                        convertedSchema["type"] = "string"
                        convertedSchema["format"] = bsonType
                    }
                }
            } else if (key === "properties") {
                convertedSchema[key] = convertObject(schema[key])
            } else if (key === "additionalProperties") {
                if (schema[key] === false) {
                    convertedSchema[key] = false
                } else if (typeof schema[key] === "object") {
                    convertedSchema[key] = convertObject(schema[key])
                }
            } else if (key === "items") {
                if (Array.isArray(schema[key])) {
                    convertedSchema[key] = schema[key].map(convertBsonType)
                } else if (typeof schema[key] === "object") {
                    convertedSchema[key] = convertBsonType(schema[key])
                }
            } else {
                convertedSchema[key] = schema[key]
            }
        }
        return convertedSchema
    }

    function convertBsonType(schema) {
        if (schema.bsonType) {
            return convertObject(schema)
        } else {
            return schema
        }
    }

    const rewritedMongoDb = convertBsonType(mongoSchema)

    return {
        "$ref": "#/definitions/mongoDB",
        "definitions": {
            "mongoDB": rewritedMongoDb
        }
    }
}

```

Explanations : 

> The implementation defines two helper functions: convertObject and convertBsonType.
> 
> The convertObject function takes an object representing a schema and returns a new object with the same schema but with certain properties converted to their Draft 7 equivalents. The function recursively applies itself to any nested objects.
> 
> The convertBsonType function takes an object representing a schema with a bsonType property and returns a new object with the same schema but with the bsonType property converted to a type property.
> 
> The main convertToDraft7 function uses the helper functions to convert the input mongoSchema to a Draft 7 schema. The converted schema is stored in the rewritedMongoDb variable and is returned in a JSON Schema Draft 7 object.

**Reply by bwgjoseph** [reply]
Thanks for the hint. 

I suppose that if you are going to re-write into draft07, then probably don't have to use `$ref` anymore and directly return the object as is.

So I will just call

```
<JSONSchemaViewer schema={convertToDraft7(schema)} />
```

Tried to do a quick hack to get it working but wasn't able to, will try it again next time.

**Reply by bwgjoseph** [reply]
Also, I think it's a good idea to include `FAQ` section

**Reply by jy95** [reply]
@bwgjoseph Indeed, the final choice is yours like how you want to fetch your schema ;) . 
You can directly return `rewritedMongoDb` in `convertToDraft7` function so you don't have to care about `$ref` :

```tsx
<JSONSchemaViewer schema={convertToDraft7(schema)} />
```

To be accurate, if I suggested  to write that function into `src/_convertToDraft7.tsx`, it is to avoid Docusaurus trying to interpreting that as a route (see [docs](https://docusaurus.io/docs/creating-pages#routing) )

> By default, any Markdown or JavaScript file starting with _ will be ignored and no routes will be created for that file (see the exclude option).

FYI : I fine-tuned the plugin (currently v1.2.1) so arrays are better handled. From what I saw in your previous issue, it might be handy ^^

**Reply by bwgjoseph** [reply]
You're too awesome. 

Thanks for all the work done, appreciate it!

I will play around with it when I can!

**Reply by jy95** [reply]
> Also, I think it's a good idea to include `FAQ` section

I added some extra documentation on the website : https://jy95.github.io/docusaurus-json-schema-plugin/ 
Let's said I was in the emoji mood ^^

![image](https://user-images.githubusercontent.com/9306961/219973093-34af07e4-4232-498d-ab7a-b6ca1ba5b4c8.png)
